mlir/lib/Analysis/SliceAnalysis.cpp

```diff
@@ -105,52 +105,47 @@ void mlir::getForwardSlice(Value root, SetVector<Operation *> *forwardSlice,
   forwardSlice->insert(v.rbegin(), v.rend());
 }
 
-static LogicalResult getBackwardSliceImpl(Operation *op,
-                                          DenseSet<Operation *> &visited,
-                                          SetVector<Operation *> *backwardSlice,
-                                          const BackwardSliceOptions &options) {
+static void getBackwardSliceImpl(Operation *op, DenseSet<Operation *> &visited,
+                                 SetVector<Operation *> *backwardSlice,
+                                 const BackwardSliceOptions &options) {
   if (!op)
-    return success();
+    return;
 
   // Evaluate whether we should keep this def.
   // This is useful in particular to implement scoping; i.e. return the
   // transitive backwardSlice in the current scope.
   if (options.filter && !options.filter(op))
-    return success();
+    return;
 
   auto processValue = [&](Value value) {
     if (auto *definingOp = value.getDefiningOp()) {
       if (backwardSlice->count(definingOp) == 0 &&
-          visited.insert(definingOp).second)
-        return getBackwardSliceImpl(definingOp, visited, backwardSlice,
-                                    options);
+          visited.insert(definingOp).second) {
+        getBackwardSliceImpl(definingOp, visited, backwardSlice, options);
+        return;
+      }
 
       visited.erase(definingOp);
-    } else if (auto blockArg = dyn_cast<BlockArgument>(value)) {
-      if (options.omitBlockArguments)
-        return success();
-
-      Block *block = blockArg.getOwner();
-      Operation *parentOp = block->getParentOp();
-      // TODO: determine whether we want to recurse backward into the other
-      // blocks of parentOp, which are not technically backward unless they flow
-      // into us. For now, just bail.
-      if (parentOp && backwardSlice->count(parentOp) == 0) {
-        if (!parentOp->hasTrait<OpTrait::IsIsolatedFromAbove>() &&
-            parentOp->getNumRegions() == 1 &&
-            parentOp->getRegion(0).hasOneBlock()) {
-          return getBackwardSliceImpl(parentOp, visited, backwardSlice,
-                                      options);
-        }
+      return;
+    }
+    auto blockArg = cast<BlockArgument>(value);
+    if (options.omitBlockArguments)
+      return;
+
+    Block *block = blockArg.getOwner();
+    Operation *parentOp = block->getParentOp();
+    // TODO: determine whether we want to recurse backward into the other
+    // blocks of parentOp, which are not technically backward unless they flow
+    // into us. For now, just bail.
+    if (parentOp && backwardSlice->count(parentOp) == 0) {
+      if (!parentOp->hasTrait<OpTrait::IsIsolatedFromAbove>() &&
+          parentOp->getNumRegions() == 1 &&
+          parentOp->getRegion(0).hasOneBlock()) {
+        getBackwardSliceImpl(parentOp, visited, backwardSlice, options);
       }
-    } else {
-      return failure();
```

Comment:
Note that this case can never happen: An SSA value is always either an OpResult or a BlockArgument.